File: Practical-5/main.py
def split_input(input_str):
    """Helper function to split comma-separated input into list of tokens"""
    return [token.strip() for token in input_str.split(",") if token.strip()]


def calculate_precision_recall(A, R):
    setA = set(A)
    setR = set(R)

    Ra = setA.intersection(setR)

    precision = len(Ra) / len(setA) if len(setA) > 0 else 0
    recall = len(Ra) / len(setR) if len(setR) > 0 else 0

    return len(Ra), precision, recall


def calculate_f_measure(precision, recall):
    return (2 * precision * recall / (precision + recall)) if (precision + recall) > 0 else 0


def calculate_e_measure(precision, recall, beta):
    if precision == 0 and recall == 0:
        return 0
    return ((1 + beta ** 2) * precision * recall) / ((beta ** 2 * precision) + recall) if ((beta ** 2 * precision) + recall) > 0 else 0


def generate_table(retrieved_docs, relevant_docs):
    print(f"{'Documents':<60} {'|Ra|':<5} {'|A|':<5} {'Precision':<10} {'Recall':<10}")
    print("-" * 95)

    results = []

    for i in range(1, len(retrieved_docs) + 1):
        current_set = retrieved_docs[:i]
        Ra_size, precision, recall = calculate_precision_recall(current_set, relevant_docs)

        results.append((precision, recall))

        docs_str = ", ".join(current_set)
        print(f"{docs_str:<60} {Ra_size:<5} {len(current_set):<5} {precision*100:<10.2f} {recall*100:<10.2f}")

    return results


def main():
    # Input retrieved documents
    retrieved_input = input("Enter the retrieved documents (comma-separated): ")
    retrieved_docs = split_input(retrieved_input)

    # Input relevant documents
    relevant_input = input("Enter the relevant documents (comma-separated): ")
    relevant_docs = split_input(relevant_input)

    # Generate table
    results = generate_table(retrieved_docs, relevant_docs)

    # Choose index j
    j = int(input(f"\nEnter value of j (0 - {len(results)-1}) to find F(j) and E(j): "))

    precision, recall = results[j]

    f1 = calculate_f_measure(precision, recall)

    e_greater = calculate_e_measure(precision, recall, 2)
    e_equal = calculate_e_measure(precision, recall, 1)
    e_less = calculate_e_measure(precision, recall, 0.5)

    print("\n---------------------------------")
    print(f"| Harmonic mean (F1) is: | {f1:.2f} |")
    print("---------------------------------\n")

    print("        ----------------------------")
    print("        |          E-Value          |")
    print("        ----------------------------")
    print("        |   b>1   |   b=1   |  b<1  |")
    print("        ----------------------------")
    print(f"        |  {e_greater:.2f}  |  {e_equal:.2f}  |  {e_less:.2f} |")
    print("        ----------------------------")


if __name__ == "__main__":
    main()

File: Practical-5/main.cpp
#include <iostream>
#include <vector>
#include <set>
#include <sstream>
#include <iomanip>
#include <cmath>
using namespace std;

// Helper function to split comma-separated input into list of tokens
vector<string> split_input(const string& input_str) {
    vector<string> tokens;
    stringstream ss(input_str);
    string token;
    while (getline(ss, token, ',')) {
        size_t start = token.find_first_not_of(" \t");
        size_t end = token.find_last_not_of(" \t");
        if (start != string::npos && end != string::npos)
            tokens.push_back(token.substr(start, end - start + 1));
    }
    return tokens;
}

// Calculate precision and recall
tuple<int, double, double> calculate_precision_recall(const vector<string>& A, const vector<string>& R) {
    set<string> setA(A.begin(), A.end());
    set<string> setR(R.begin(), R.end());

    int Ra = 0;
    for (const auto& doc : setA) {
        if (setR.count(doc)) Ra++;
    }

    double precision = setA.size() > 0 ? (double)Ra / setA.size() : 0;
    double recall = setR.size() > 0 ? (double)Ra / setR.size() : 0;

    return {Ra, precision, recall};
}

// Calculate F-measure
double calculate_f_measure(double precision, double recall) {
    return (precision + recall) > 0 ? (2 * precision * recall) / (precision + recall) : 0;
}

// Calculate E-measure
double calculate_e_measure(double precision, double recall, double beta) {
    if (precision == 0 && recall == 0) return 0;
    double numerator = (1 + beta * beta) * precision * recall;
    double denominator = (beta * beta * precision) + recall;
    return denominator > 0 ? numerator / denominator : 0;
}

// Generate table and collect precision/recall pairs
vector<pair<double, double>> generate_table(const vector<string>& retrieved_docs, const vector<string>& relevant_docs) {
    cout << left << setw(60) << "Documents"
         << setw(5) << "|Ra|"
         << setw(5) << "|A|"
         << setw(10) << "Precision"
         << setw(10) << "Recall" << endl;
    cout << string(95, '-') << endl;

    vector<pair<double, double>> results;

    for (size_t i = 1; i <= retrieved_docs.size(); ++i) {
        vector<string> current_set(retrieved_docs.begin(), retrieved_docs.begin() + i);
        int Ra_size;
        double precision, recall;
        tie(Ra_size, precision, recall) = calculate_precision_recall(current_set, relevant_docs);

        results.push_back({precision, recall});

        string docs_str;
        for (size_t j = 0; j < i; ++j) {
            docs_str += retrieved_docs[j];
            if (j < i - 1) docs_str += ", ";
        }

        cout << left << setw(60) << docs_str
             << setw(5) << Ra_size
             << setw(5) << i
             << setw(10) << fixed << setprecision(2) << precision * 100
             << setw(10) << fixed << setprecision(2) << recall * 100 << endl;
    }

    return results;
}

int main() {
    string retrieved_input, relevant_input;

    // Input retrieved documents
    cout << "Enter the retrieved documents (comma-separated): ";
    getline(cin, retrieved_input);
    vector<string> retrieved_docs = split_input(retrieved_input);

    // Input relevant documents
    cout << "Enter the relevant documents (comma-separated): ";
    getline(cin, relevant_input);
    vector<string> relevant_docs = split_input(relevant_input);

    // Generate table
    vector<pair<double, double>> results = generate_table(retrieved_docs, relevant_docs);

    // Choose index j
    int j;
    cout << "\nEnter value of j (0 - " << results.size() - 1 << ") to find F(j) and E(j): ";
    cin >> j;

    double precision = results[j].first;
    double recall = results[j].second;

    double f1 = calculate_f_measure(precision, recall);

    double e_greater = calculate_e_measure(precision, recall, 2);
    double e_equal = calculate_e_measure(precision, recall, 1);
    double e_less = calculate_e_measure(precision, recall, 0.5);

    cout << "\n---------------------------------\n";
    cout << "| Harmonic mean (F1) is: | " << fixed << setprecision(2) << f1 << " |\n";
    cout << "---------------------------------\n\n";

    cout << "        ----------------------------\n";
    cout << "        |          E-Value          |\n";
    cout << "        ----------------------------\n";
    cout << "        |   b>1   |   b=1   |  b<1  |\n";
    cout << "        ----------------------------\n";
    cout << "        |  " << fixed << setprecision(2) << e_greater
         << "  |  " << e_equal
         << "  |  " << e_less << " |\n";
    cout << "        ----------------------------\n";

    return 0;
}